File: Engine/src/main/java/org/etieskrill/engine/graphics/model/loader/MeshProcessor.java
package org.etieskrill.engine.graphics.model.loader;

import org.etieskrill.engine.graphics.model.*;
import org.joml.Vector2f;
import org.joml.Vector2fc;
import org.joml.Vector3f;
import org.joml.Vector3fc;
import org.joml.primitives.AABBf;
import org.lwjgl.BufferUtils;
import org.lwjgl.PointerBuffer;
import org.lwjgl.assimp.AIFace;
import org.lwjgl.assimp.AIMesh;
import org.lwjgl.assimp.AIScene;
import org.lwjgl.assimp.AIVector3D;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import static org.etieskrill.engine.graphics.model.loader.AnimationLoader.getBones;
import static org.lwjgl.assimp.Assimp.*;
import static org.lwjgl.util.meshoptimizer.MeshOptimizer.meshopt_simplify;

public class MeshProcessor {

    private static final Logger logger = LoggerFactory.getLogger(MeshProcessor.class);

    static void loadMeshes(AIScene scene, Model.Builder builder) {
        PointerBuffer meshBuffer = scene.mMeshes();
        if (meshBuffer == null) return;

        List<Mesh> meshes = Stream
                .generate(meshBuffer::get)
                .limit(scene.mNumMeshes())
                .map(AIMesh::create)
                .map(aiMesh -> processMesh(aiMesh, builder.getMaterials()))
                .toList();
        builder.getMeshes().addAll(meshes);
        builder.getBones().addAll(meshes.stream()
                .flatMap(mesh -> mesh.getBones().stream())
                .distinct()
                .toList());
    }

    private static Mesh processMesh(AIMesh aiMesh, List<Material> materials) {
        //TODO add AABB loading but oh FUCK each goddamn mesh has a separate one FFS

        int numVertices = aiMesh.mNumVertices();
        List<Vector3fc> positions = new ArrayList<>(numVertices);
        aiMesh.mVertices().forEach(vertex -> positions.add(new Vector3f(vertex.x(), vertex.y(), vertex.z())));

        List<Vector3fc> normals = new ArrayList<>(numVertices);
        if (aiMesh.mNormals() != null)
            aiMesh.mNormals().forEach(normal -> normals.add(new Vector3f(normal.x(), normal.y(), normal.z())));

        List<Vector2fc> texCoords = new ArrayList<>(numVertices);
        if (aiMesh.mTextureCoords(0) != null)
            aiMesh.mTextureCoords(0).forEach(texCoord -> texCoords.add(new Vector2f(texCoord.x(), texCoord.y())));

        List<Vector3fc> tangents = new ArrayList<>(numVertices);
        List<Vector3fc> biTangents = new ArrayList<>(numVertices);
        if (aiMesh.mTangents() != null) {
            aiMesh.mTangents().forEach(tangent -> tangents.add(new Vector3f(tangent.x(), tangent.y(), tangent.z())));
            aiMesh.mBitangents().forEach(bitangent -> biTangents.add(new Vector3f(bitangent.x(), bitangent.y(), bitangent.z())));
        }

        List<Vertex.VertexBuilder> vertexBuilders = new ArrayList<>(numVertices);
        for (int i = 0; i < aiMesh.mNumVertices(); i++) vertexBuilders.add(Vertex.builder(positions.get(i)));
        if (!normals.isEmpty()) {
            for (int i = 0; i < aiMesh.mNumVertices(); i++)
                vertexBuilders.get(i).normal(normals.get(i));
        }
        if (!texCoords.isEmpty()) {
            for (int i = 0; i < aiMesh.mNumVertices(); i++)
                vertexBuilders.get(i).textureCoords(texCoords.get(i));
        }
        if (!tangents.isEmpty()) {
            for (int i = 0; i < aiMesh.mNumVertices(); i++)
                vertexBuilders.get(i).tangent(tangents.get(i)).biTangent(biTangents.get(i));
        }

        //three because a face is usually a triangle, but this list is discarded at the first opportunity a/w
        //TODO add loader versions which transmit the minimal amount of data (shorts for indices, smaller vectors)
        boolean calcTangentsWarning = false;
        List<Integer> indices = new ArrayList<>(aiMesh.mNumFaces() * 3);
        for (int i = 0; i < aiMesh.mNumFaces(); i++) {
            AIFace face = aiMesh.mFaces().get(i);
            IntBuffer buffer = face.mIndices();

//            boolean calcTangents = face.mNumIndices() == 3;
//            if (!calcTangentsWarning && !calcTangents) {
//                logger.warn("Primitive type with {} vertices will not have tangents calculated", face.mNumIndices());
//                calcTangentsWarning = true;
//            }
//
//            Vertex.Builder vertex1 = null, vertex2 = null, vertex3 = null;
            for (int j = 0; j < face.mNumIndices(); j++) {
                int index = buffer.get();
                indices.add(index);
//
//                if (!calcTangents) continue;
//                switch (j) {
//                    case 0 -> vertex1 = vertexBuilders.get(index);
//                    case 1 -> vertex2 = vertexBuilders.get(index);
//                    case 2 -> vertex3 = vertexBuilders.get(index);
//                }
            }
//
//            Vector3fc edge1 = vertex2.position().sub(vertex1.position(), new Vector3f());
//            Vector3fc edge2 = vertex3.position().sub(vertex1.position(), new Vector3f());
//            Vector2fc deltaTex1 = vertex2.textureCoords().sub(vertex1.textureCoords(), new Vector2f());
//            Vector2fc deltaTex2 = vertex3.textureCoords().sub(vertex1.textureCoords(), new Vector2f());
//
//            Vector3f tangent = new Vector3f();
//            Vector3f biTangent = new Vector3f();
//
//            float f = 1.0f / (deltaTex1.x() * deltaTex2.y() - deltaTex2.x() * deltaTex1.y());
//
//            tangent.x = f * (deltaTex2.y() * edge1.x() - deltaTex1.y() * edge2.x());
//            tangent.y = f * (deltaTex2.y() * edge1.y() - deltaTex1.y() * edge2.y());
//            tangent.z = f * (deltaTex2.y() * edge1.z() - deltaTex1.y() * edge2.z());
//
//            biTangent.x = f * (-deltaTex2.x() * edge1.x() + deltaTex1.x() * edge2.x());
//            biTangent.y = f * (-deltaTex2.x() * edge1.y() + deltaTex1.x() * edge2.y());
//            biTangent.z = f * (-deltaTex2.x() * edge1.z() + deltaTex1.x() * edge2.z());
//
//            vertex1.tangent(tangent).biTangent(biTangent);
//            vertex2.tangent(tangent).biTangent(biTangent);
//            vertex3.tangent(tangent).biTangent(biTangent);
        }

        List<Bone> bones = getBones(aiMesh, vertexBuilders);

        List<Vertex> vertices = vertexBuilders.stream()
                .map(Vertex.VertexBuilder::build)
                .toList();

        AIVector3D min = aiMesh.mAABB().mMin();
        AIVector3D max = aiMesh.mAABB().mMax();
        AABBf boundingBox = new AABBf(min.x(), min.y(), min.z(), max.x(), max.y(), max.z());

        if (boundingBox.getSize(new Vector3f()).equals(0, 0, 0)) {
            boundingBox = calculateBoundingBox(vertices);
        }

        Material material = materials.get(aiMesh.mMaterialIndex());

        Mesh.DrawMode drawMode = switch (aiMesh.mPrimitiveTypes()) {
            case aiPrimitiveType_POINT -> Mesh.DrawMode.POINTS;
            case aiPrimitiveType_LINE -> Mesh.DrawMode.LINES;
            case aiPrimitiveType_TRIANGLE -> Mesh.DrawMode.TRIANGLES;
            default -> {
                logger.warn("Cannot draw primitives of type 0x{}, using default of TRIANGLES",
                        Integer.toHexString(aiMesh.mPrimitiveTypes()));
                yield Mesh.DrawMode.TRIANGLES;
            }
        };

        Mesh ret = MeshLoader.loadToVAO(vertices, indices, material, bones, boundingBox, drawMode);

        logger.trace("Loaded mesh with {} vertices and {} indices, {} normals, {} uv coordinates", vertices.size(),
                indices.size(), !normals.isEmpty() ? "with" : "without", !texCoords.isEmpty() ? "with" : "without");
        return ret;
    }

    private static AABBf calculateBoundingBox(List<Vertex> vertices) {
        AABBf boundingBox = new AABBf();

        //for models as complicated as the skeleton, a stream variant presents a slight improvement in performance,
        //measurable in the single milliseconds, which is not worth the cost of initialising a stream for a model with
        //very few vertices
        for (Vertex vertex : vertices) {
            boundingBox.union(vertex.getPosition());
        }

        return boundingBox;
    }

    public static void optimiseMesh(Mesh mesh, int targetIndexCount, float maxDeformation) {
        if (!mesh.getVao().isIndexed()) {
            throw new IllegalArgumentException("Can only optimise indexed meshes"); //TODO workaround: temporary 1-1 vertex-index buffer? duplicate vertices did make the algorithm shit itself tho, so probably do an actual index run using... assimp? can it even do that?
        }

        ByteBuffer vertexData = mesh.getVao().getVertexBuffer().getData();
        IntBuffer indexData = mesh.getVao().getIndexBuffer().getData().asIntBuffer();

        ByteBuffer newIndexData = BufferUtils.createByteBuffer(Integer.BYTES * indexData.capacity());

        FloatBuffer errorBuffer = BufferUtils.createFloatBuffer(1);
        int vertexBytes = VertexAccessor.getInstance().getElementByteSize();
        long numIndices = meshopt_simplify(newIndexData.asIntBuffer(), indexData, vertexData.asFloatBuffer(),
                vertexData.capacity() / vertexBytes, vertexBytes,
                targetIndexCount, maxDeformation, 0, errorBuffer);
        //TODO compress vertex buffer using new indices

        logger.trace("Original mesh has {} vertices and {} indices",
                vertexData.capacity() / vertexBytes, indexData.capacity());
        logger.trace("Optimised mesh has {} indices and a deformation of {}% (max {}%)", numIndices,
                "%5.3f".formatted(100 * errorBuffer.get()), "%5.3f".formatted(100 * maxDeformation));
        logger.debug("Mesh was compressed by a factor of {}",
                "%4.1f".formatted((float) indexData.capacity() / numIndices));

        mesh.getVao().getIndexBuffer().setData(newIndexData);
        mesh.setNumIndices((int) numIndices);
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/graphics/model/loader/MeshProcessor.kt
package org.etieskrill.engine.graphics.model.loader

import io.github.oshai.kotlinlogging.KotlinLogging
import org.etieskrill.engine.graphics.model.Bone
import org.etieskrill.engine.graphics.model.Material
import org.etieskrill.engine.graphics.model.Mesh
import org.etieskrill.engine.graphics.model.Vertex
import org.joml.Vector2f
import org.joml.Vector3f
import org.joml.Vector4f
import org.joml.Vector4i
import org.joml.primitives.AABBf
import org.lwjgl.BufferUtils
import org.lwjgl.assimp.AIMesh
import org.lwjgl.assimp.AIScene
import org.lwjgl.assimp.Assimp.*
import org.lwjgl.util.meshoptimizer.MeshOptimizer.meshopt_simplify

private val logger = KotlinLogging.logger {}

internal fun loadMeshes(
    scene: AIScene,
    materials: List<Material>,
    meshes: MutableList<Mesh>,
    bones: MutableList<Bone>
) {
    val meshBuffer = scene.mMeshes() ?: return

    val newMeshes = generateSequence { AIMesh.create(meshBuffer.get()) }
        .take(scene.mNumMeshes())
        .map { processMesh(it, materials) }
        .toList()

    meshes += newMeshes
    bones += newMeshes.flatMap { it.bones ?: emptyList() }.distinct()
}

@OptIn(ExperimentalStdlibApi::class)
private fun processMesh(aiMesh: AIMesh, materials: List<Material>): Mesh {
    val positions = aiMesh.mVertices()
    val normals = aiMesh.mNormals()
    val texCoords = aiMesh.mTextureCoords(0)
    val tangents = aiMesh.mTangents()
    val biTangents = aiMesh.mBitangents()

    val (bones, boneWeights) = getBones(aiMesh)

    val vertices = List(aiMesh.mNumVertices()) { i ->
        Vertex(
            position = positions.get().let { Vector3f(it.x(), it.y(), it.z()) },
            normal = normals?.get()?.let { Vector3f(it.x(), it.y(), it.z()) },
            textureCoords = texCoords?.get()?.let { Vector2f(it.x(), it.y()) },
            tangent = tangents?.get()?.let { Vector3f(it.x(), it.y(), it.z()) },
            biTangent = biTangents?.get()?.let { Vector3f(it.x(), it.y(), it.z()) },
            bones = boneWeights?.get(i)?.bones ?: Vector4i(-1),
            boneWeights = boneWeights?.get(i)?.boneWeights ?: Vector4f(0f)
        )
    }

    val indices = mutableListOf<Int>()
    val faceBuffer = aiMesh.mFaces()
    repeat(aiMesh.mNumFaces()) {
        val face = faceBuffer.get()
        val indexBuffer = face.mIndices()
        repeat(face.mNumIndices()) {
            indices += indexBuffer.get()
        }
    }

//    val min = aiMesh.mAABB().mMin()
//    val max = aiMesh.mAABB().mMax()
//    val boundingBox = AABBf(min.x(), min.y(), min.z(), max.x(), max.y(), max.z())
    val boundingBox = calculateBoundingBox(vertices)

    val material = materials[aiMesh.mMaterialIndex()]

    val primitiveType = aiMesh.mPrimitiveTypes()
    if (primitiveType and aiPrimitiveType_NGONEncodingFlag > 0) {
        logger.info { "Mesh contains n-gon encoding, which is not explicitly supported; if a mesh looks weird, this may be the reason" } //FIXME i dunnot if this even has any impact
    }
    val drawMode = when (primitiveType and aiPrimitiveType_NGONEncodingFlag.inv()) {
        aiPrimitiveType_POINT -> Mesh.DrawMode.POINTS
        aiPrimitiveType_LINE -> Mesh.DrawMode.LINES
        aiPrimitiveType_TRIANGLE -> Mesh.DrawMode.TRIANGLES
        else -> error("Unsupported primitive type: 0x${primitiveType.toHexString()}")
    }

    val mesh = loadToVAO(vertices, indices, material, bones, boundingBox, drawMode)

    logger.trace { "Loaded mesh with ${vertices.size} vertices and ${indices.size} indices" }
    return mesh
}

fun calculateBoundingBox(vertices: List<Vertex>) = AABBf().apply { vertices.forEach { union(it.position) } }

fun optimiseMesh(mesh: Mesh, targetIndexCount: Int, maxDeformation: Float) {
    val vertexData = mesh.vao.vertexBuffer.getData()
    //TODO workaround: temporary 1-1 vertex-index buffer? duplicate vertices did make the algorithm shit itself tho, so probably do an actual index run using... assimp? can it even do that?
    val indexData = mesh.vao.indexBuffer?.getData()?.asIntBuffer() ?: error("Can only optimise indexed meshes")

    val newIndexData = BufferUtils.createByteBuffer(Int.SIZE_BYTES * indexData.capacity())

    val errorBuffer = BufferUtils.createFloatBuffer(1)

    val vertexBytes = Vertex.Companion.Accessor.elementByteSize.toLong()
    val numIndices = meshopt_simplify(
        newIndexData.asIntBuffer(), indexData, vertexData.asFloatBuffer(),
        vertexData.capacity() / vertexBytes, vertexBytes,
        targetIndexCount.toLong(), maxDeformation, 0, errorBuffer
    )
    //TODO compress vertex buffer using new indices

    logger.trace { "Original mesh has ${vertexData.capacity() / vertexBytes} vertices and ${indexData.capacity()} indices" }
    logger.trace {
        "Optimised mesh has %d indices and a deformation of %5.3f%% (max %5.3f%%)"
            .format(numIndices, 100 * errorBuffer.get(), 100 * maxDeformation)
    }
    logger.debug { "Mesh was compressed by a factor of %4.1f".format(indexData.capacity().toFloat() / numIndices) }

    mesh.vao.indexBuffer.setData(newIndexData)
}
